### ISSUE TITLE
Base.HWReal on Apple silicon

### ISSUE THREAD

**Issue by ctkelley** [JuliaLang/julia#50804]
Why is Float16 missing from Base.HWReal on Apple silicon? On my M2 Pro, Julia 1.10 I get
```
julia> Base.HWReal
Union{Float32, Float64, Int16, Int32, Int64, Int8, UInt16, UInt32, UInt64, UInt8}
```

**Comment by vchuravy** [comment]
There is a bit of a challenge here. HWReal is per architecture, and `Float16` is per micro-architecture.

x-ref: #46499 #43327 #40216

**Comment by ctkelley** [comment]
OK. Some useful packages query ```Base.HWReal```. ```LoopVecrtorization.jl``` is one example.

**Comment by vchuravy** [comment]
Yeah we would need to have a `has_float16` predicate like `has_fma`

**Comment by giordano** [comment]
Fixed by #55929.
